metadataclient/mds.py

```diff
@@ -0,0 +1,259 @@
+                                     self._RUN_START_CACHE)
+
+    def run_stop_given_uid(self, uid):
+        uid = self.doc_or_uid_to_uid(uid)
+        try:
+            return self._RUN_STOP_CACHE[uid]
+        except KeyError:
+            pass
+        params = self.queryfactory(query={'uid': uid},
+                                   signature='run_start_given_uid')
+        response = self._get(self._rstop_url, params=params)
+        return self._cache_run_stop(run_stop=response,
+                                    self._RUN_STOP_CACHE)
+
+    def descriptor_given_uid(self, uid):
+        uid = self.doc_or_uid_to_uid(uid)
+        try:
+            return self.DESCRIPTOR_CACHE[uid]
+        except KeyError:
+            pass
+        params = self.queryfactory(query={'uid': uid},
+                                   signature='run_start_given_uid')
+        response = self._get(self._desc_url, params=params)
+        return self._cache_descriptor(descriptor=response,
+                                      self._DESCRIPTOR_CACHE)
+
+    def descriptors_by_start(run_start):
+        rstart_uid = self.doc_or_uid_to_uid(run_start)
+        params = self.queryfactor(query={'run_start': rstart_uid},
+                             signature='descriptors_by_start')
+        self._get(self._desc_url, params=params)
+        return self._cache_descriptor(descriptor=response,
+                                      self._DESCRIPTOR_CACHE)
+
+    def stop_by_start(self, run_start):
+        uid = self.doc_or_uid_to_uid()
+        params = self.queryfactory(query={'run_start': uid},
+                                   signature='stop_by_start')
+        response = self._get(self._rstop_url, params=params)
+        return self._cache_run_stop(response, self._RUN_STOP_CACHE)
+
+    def get_events_generator(descriptor, convert_arrays=True):
+        pass
+
+    def get_events_table(descriptor):
+        pass
+
+    def find():
+        pass
+
+    def find_last():
+        pass
+
+class MDS(MDSRO):
+   _INS_METHODS = {'start': 'insert_run_start',
+                    'stop': 'insert_run_stop',
+                    'descriptor': 'insert_descriptor',
+                    'event': 'insert_event',
+                    'bulk_events': 'bulk_insert_events'}
+
+    def datafactory(self, data, signature):
+        return dict(data=data, signature=signature)
+
+    def _post(self, url, data)
+        r = request.post(url, json.dumps(data))
+        r.raise_for_status()
+        return r.json()
+
+    def insert(self):
+        pass
+
+    def _check_for_custom(self, kdict):
+        if 'custom' in kdict:
+            warnings.warn("Custom is a deprecated field")
+            custom = kdict.pop('custom')
+            if any(k in kdict for k in custom):
+                raise TypeError("Duplicate keys in kwargs and custom")
+            kdict.update(custom)
+        return kdict
+
+    def insert_run_start(self, time, uid, **kwargs):
```

Comment:
Until we sort out the JSON stuff (which, as you know, I think we can do *after* this is initially deployed) let's make these raise a `NotImplementedError` to be sure no one attaches this to a RunEngine.